File: java/72.java
class Solution {
    private int min(int i, int j) {
        return (i < j) ? i : j;
    }

    private int min3(int i, int j, int k) {
        return min(i, min(j, k));
    }

    public int minDistance(String sw1, String sw2) {
        int n1 = sw1.length(), n2 = sw2.length();
        if (n1 == 0)
            return n2;
        if (n2 == 0)
            return n1;
        char w1[] = sw1.toCharArray();
        char w2[] = sw2.toCharArray();
        int dp[][] = new int[n1][n2];
        return find(w1, 0, w2, 0, dp);
    }

    private int find(char s1[], int i, char s2[], int j, int dp[][]) {
        if (i == s1.length)
            return s2.length - j;
        if (j == s2.length)
            return s1.length - i;
        if (dp[i][j] != 0)
            return dp[i][j];
        int res;
        if (s1[i] == s2[j])
            res = find(s1, i + 1, s2, j + 1, dp);
        else
            res = min3(find(s1, i + 1, s2, j + 1, dp),
                    find(s1, i + 1, s2, j, dp), find(s1, i, s2, j + 1, dp)) + 1;
        dp[i][j] = res;
        return res;
    }
}

File: python3/72.py
class Solution:
    def minDistance(self, w1: str, w2: str) -> int:
        def find(s1: str, i: int, s2: str, j: int, dp: List[List[int]]):
            if i == len(s1):
                return len(s2) - j
            if j == len(s2):
                return len(s1) - i
            if dp[i][j] != -1:
                return dp[i][j]
            res = -1
            if s1[i] == s2[j]:
                res = find(s1, i + 1, s2, j + 1, dp)
            else:
                res = (
                    min(
                        find(s1, i + 1, s2, j + 1, dp),
                        min(find(s1, i + 1, s2, j, dp), find(s1, i, s2, j + 1, dp)),
                    )
                    + 1
                )
            dp[i][j] = res
            return res

        n1 = len(w1)
        n2 = len(w2)
        dp = [[-1] * n2 for _ in range(n1)]
        return find(w1, 0, w2, 0, dp)
